git diff --git a/my-app/src/App.tsx b/my-app/src/App.tsx
index d96bc27..083951c 100644
--- a/my-app/src/App.tsx
+++ b/my-app/src/App.tsx
@@ -4,34 +4,56 @@ import { DisplayAmount } from "./Components/DisplayAmount";
 import { DoughShelf } from "./Components/DoughShelf";
 import { Oven } from "./Components/Oven";
 import { GlobalStyle } from "./GlobalStyles";
+import { Dough } from "./types";
 
 function App() {
   const [flourAmount, setFlourAmount] = useState<number>(40);
   const [progresDought, setProgresDought] = useState<number>(0);
   const [isMakingDought, setisMakingDought] = useState<boolean>(false);
   const [doughtAmount, setDoughtAmount] = useState<number>(0);
-  const [doughtArray, setDoughtArray] = useState<number[]>([]);
+  const [doughtArray, setDoughtArray] = useState<Dough[]>([]);
   const [rowCookie, setRowCookie] = useState<number>(0);
   const [cookiesInOven, setCookiesInOven] = useState<number>(0);
 
+  const makeCookie = (doughId: number) => {
+    setDoughtArray((old) =>
+      old
+        .map((dough) =>
+          dough.id !== doughId ? dough : { ...dough, size: dough.size - 10 }
+        )
+        .filter((dough) => dough.size > 0)
+    );
+    setRowCookie((old) => old + 1);
+  };
+
   useEffect(() => {
     if (flourAmount < 10) {
       setisMakingDought(false);
       return;
     }
+    let timeoutId = 0;
     if (isMakingDought) {
       if (progresDought === 100) {
         setProgresDought(0);
-        setDoughtAmount(doughtAmount + 1);
-        setFlourAmount(flourAmount - 10);
-        setDoughtArray((doughtArray) => [...doughtArray, 1]);
+        setDoughtAmount((doughtAmount) => doughtAmount + 1);
+        setFlourAmount((flourAmount) => flourAmount - 10);
+        setDoughtArray((doughtArray) => [
+          ...doughtArray,
+          { id: Date.now(), size: 100 },
+        ]);
 
         return;
       }
-      setTimeout(() => {
+      timeoutId = window.setTimeout(() => {
         return setProgresDought(progresDought + 1);
       }, 10);
     }
+
+    return () => {
+      if (timeoutId) {
+        clearTimeout(timeoutId);
+      }
+    };
   }, [progresDought, isMakingDought]);
 
   return (
@@ -62,18 +84,13 @@ function App() {
         nameAmount={"Ilość surowych ciasteczek:"}
         type={"szt"}
       />
-      {doughtArray.map((e, index, array) => {
+      {doughtArray.map((element) => {
         return (
-          <>
-            <DoughShelf
-              key={index}
-              setState={setRowCookie}
-              rowCookie={rowCookie}
-              setDoughtArray={setDoughtArray}
-              doughtArray={doughtArray}
-              index={index}
-            />
-          </>
+          <DoughShelf
+            key={element.id}
+            size={element.size}
+            onDoughClick={() => makeCookie(element.id)}
+          />
         );
       })}
 
diff --git a/my-app/src/Components/DoughShelf.tsx b/my-app/src/Components/DoughShelf.tsx
index b3321f1..7decef8 100644
--- a/my-app/src/Components/DoughShelf.tsx
+++ b/my-app/src/Components/DoughShelf.tsx
@@ -1,5 +1,4 @@
-import { stat } from "fs";
-import React, { useState } from "react";
+import React from "react";
 import styled from "styled-components";
 
 const Ball = styled.div`
@@ -8,43 +7,15 @@ const Ball = styled.div`
 `;
 
 interface Props {
-  setState: React.Dispatch<React.SetStateAction<any>>;
-  rowCookie: number;
-  setDoughtArray: React.Dispatch<React.SetStateAction<any>>;
-  doughtArray: number[];
-  index: number;
-}
-
-interface ArrayProps {
-  index: number;
-  setDoughtArray: React.Dispatch<React.SetStateAction<any>>;
-  doughtArray: number[];
+  onDoughClick: () => void;
   size: number;
 }
 
-export const DoughShelf: React.FC<Props> = ({
-  setState,
-  rowCookie,
-  setDoughtArray,
-  doughtArray,
-  index,
-}) => {
-  const [size, setSize] = useState(100);
-
-  function deleteItem(props: ArrayProps) {
-    if (size === 10) {
-      return setDoughtArray(doughtArray.splice(index, 1));
-    } else return;
-  }
-
+export const DoughShelf: React.FC<Props> = ({ onDoughClick, size }) => {
   return (
     <div>
       <Ball
-        onClick={(e) => {
-          setSize(size - 10);
-          setState(rowCookie + 1);
-          deleteItem({ index, doughtArray, size, setDoughtArray });
-        }}
+        onClick={onDoughClick}
         style={{
           width: size,
           height: size,
diff --git a/my-app/tsconfig.json b/my-app/tsconfig.json
index a273b0c..9d379a3 100644
--- a/my-app/tsconfig.json
+++ b/my-app/tsconfig.json
@@ -1,11 +1,7 @@
 {
   "compilerOptions": {
     "target": "es5",
-    "lib": [
-      "dom",
-      "dom.iterable",
-      "esnext"
-    ],
+    "lib": ["dom", "dom.iterable", "esnext"],
     "allowJs": true,
     "skipLibCheck": true,
     "esModuleInterop": true,
@@ -20,7 +16,5 @@
     "noEmit": true,
     "jsx": "react-jsx"
   },
-  "include": [
-    "src"
-  ]
+  "include": ["src"]
 }
